### PR TITLE
bz1320302 - docker --cgroup-parent=X

### PR DESCRIPTION
**Pull request by edsantiago**
Test that --cgroup-parent=X and X-Y behave as expected.
Test valid only on docker >= 1.9.1-25.el7 and >= 1.10.

Signed-off-by: Ed Santiago <santiago@redhat.com>

### CODE REVIEW

File: config_defaults/subtests/docker_cli/run_cgroups.ini [hunk 1]
```diff
@@ -3,7 +3,7 @@
 cgroup_path = /sys/fs/cgroup/memory/system.slice/docker
 #: which key is used in this test
 cgroup_key_value = memory.limit_in_bytes
-subsubtests = cpu_positive,cpu_zero,memory_positive,memory_no_cgroup,cpu_none,memory_negative,cpu_overflow
+subsubtests = cpu_positive,cpu_zero,memory_positive,memory_no_cgroup,cpu_none,memory_negative,cpu_overflow,cgroup_parent_invalid_name,cgroup_parent_path,cgroup_parent_path_with_hyphens
```
Comment (edsantiago) [inline config_defaults/subtests/docker_cli/run_cgroups.ini]: I would recommend committing this addition, but disabling it in CI until the appropriate versions of docker wend their way into our testing. Is there an automated way to do that?

File: subtests/docker_cli/run_cgroups/run_cgroups.py [hunk 2]
```diff
@@ -55,3 +56,42 @@ class memory_no_cgroup(memory_base):
...
+    """
+    Invalid value for --cgroup-parent option
+    """
+
+    def initialize(self):
+        super(cgroup_parent_invalid_name, self).initialize()
+        self._setup("/foobar",
+                    "",
+                    "Error response from daemon: Cannot start container {cid}:"
+                    " [8] System error: Invalid slice name /foobar")
```
Comment (edsantiago) [inline subtests/docker_cli/run_cgroups/run_cgroups.py]: I know the traditional way is to define these in a .ini file, and leave the class empty, but I just find that too painful to understand. How do you do it? Do you have tools or IDE or something that can superimpose the .ini settings with the code, making it possible to understand in one glance? What am I missing?

Comment (cevich) [inline subtests/docker_cli/run_cgroups/run_cgroups.py]: I don't believe I wrote this test, or at least haven't looked at it in a few decades or three.  The code is normally all in the parent class, so the empty sub-subtest class is really just there to create the name.  The name needs to be held so it can be found/loaded by the ``SubsubtestCaller()`` (or w/e it's name is) in the subtest API.  No fancy IDE's just two side-by-side terminal windows :grinning:

File: subtests/docker_cli/run_cgroups/cgroup_parent.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
+"""
```
Comment (cevich) [inline subtests/docker_cli/run_cgroups/cgroup_parent.py]: BTW, don't feel compelled to add new sub-subtests into existing subtest modules if they don't fit or it would be ugly.  Feel free to create new top-level subtests as needed.

Comment (edsantiago) [inline subtests/docker_cli/run_cgroups/cgroup_parent.py]: Thanks for the reality check. I started under run_cgroups because I thought it fit in well with that subtest; now it's changed enough that it makes sense to split out. Would run_cgroup_parent be an acceptable name?

File: subtests/docker_cli/run_cgroups/run_cgroups.py (hunk 2 above)
Comment (cevich) [inline subtests/docker_cli/run_cgroups/run_cgroups.py]: Reading more, no I think what you have here with the call to _setup() is fine.  There are many other tests that rely on shared initialization functions and overriding ``initialize()``.  No prob. doing it this way.

File: subtests/docker_cli/run_cgroups/cgroup_parent.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+from autotest.client import utils
+from dockertest.dockercmd import DockerCmd
+from dockertest.images import DockerImage
+from dockertest.output import OutputGood
+from dockertest.xceptions import DockerTestError, DockerTestFail
+from cgroups_base import cgroups_base
+
+
+class cgroup_parent_base(cgroups_base):
+    def _setup(self, cg_parent, expected_path, expected_stderr):
```
Comment (cevich) [inline subtests/docker_cli/run_cgroups/cgroup_parent.py]: pylint might get mad there's no space between class/def.  I'll let it be the judge though :smile:

File: config_defaults/subtests/docker_cli/run_cgroups.ini (hunk 1 above)
Comment (edsantiago) [inline config_defaults/subtests/docker_cli/run_cgroups.ini]: Ugh. No clean solution, then. OK, back to the drawing board.

File: subtests/docker_cli/run_cgroups/cgroup_parent.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        stderr_exp = self.sub_stuff['expected_stderr'].format(**self.sub_stuff)
+
+        # Check stderr first: that way if we're not expecting an error, but
+        # get one, our user has a better chance of seeing something useful.
+        # (Note: Running docker with -D (debug) produces lots of cruft:
+        #    time="...." level=debug msg="unwanted stuff"
+        # Strip them out.)
+        stderr = "\n".join([line
+                            for line in cmdresult.stderr.strip().split("\n")
+                            if not line.startswith('time="')])
```
Comment (cevich) [inline subtests/docker_cli/run_cgroups/cgroup_parent.py]: Just wait and see, they'll change the format in the next version, then three more times after that..  Feel free to just examine ``self.config['docker_options']`` and strip out any '-D' since it impacts the test so greatly.  TBH, I'm not sure I even like the '-D' in there by default.  It's not really ever been that useful, even when things do go south.  I'd also support nuking it from ``defaults.ini``, but I suppose you never know when it might get switched back on.  Uggg, it's never easy.

Maybe stderr only needs to be examined if the exit code is non-zero?

Hmmm.

Failing to think how this could be made simpler / avoided entirely.

File: subtests/docker_cli/run_cgroups/cgroup_parent.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+                                         % line)
+                self.failif_ne(m.group(3), path_exp, "cgroup path for %s:%s"
+                               % (m.group(1), m.group(2)))
+                found_match = True
+
+            # Must find at least one matching cgroup line
+            if not found_match:
+                raise DockerTestFail("No output from cgroups")
+        else:
+            self.failif_ne(stdout, '', "unexpected output on stdout")
```
Comment (cevich) [inline subtests/docker_cli/run_cgroups/cgroup_parent.py]: Nice, well written, easy to read.

File: subtests/docker_cli/run_cgroups/run_cgroups.py [hunk 7]
```diff
@@ -55,3 +56,41 @@ class memory_no_cgroup(memory_base):
...
+class cgroup_parent_path_with_hyphens(cgroup_parent_base):
+    """
+    Hyphens result in a more complicated path.
+    """
+
+    def initialize(self):
+        super(cgroup_parent_path_with_hyphens, self).initialize()
+        self._setup("{rand1}-{rand2}.slice",
+                    "/{rand1}.slice/{rand1}-{rand2}.slice/docker-{cid}.scope",
+                    "")
```
Comment (cevich) [inline subtests/docker_cli/run_cgroups/run_cgroups.py]: I think this is fine, I agree it's a little hard to follow and keep straight, but it's a rush-request and plenty of other tests do it this way.  If you want to try a different approach on a less-urgent test, please feel free, I'm happy to take a look.

File: subtests/docker_cli/run_cgroups/cgroup_parent.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+        OutputGood(cmdresult, ignore_error=False)
+
+        # Expected output may include the container ID; allow caller
+        # to use "{cid}", which we replace with actual one for this run.
+        cid = open(self.sub_stuff['cidfile'], 'rb').read().strip()
+        if len(cid) < 12:
+            raise DockerTestError("bad cid (length < 12) in --cidfile")
+        self.sub_stuff['cid'] = cid
+        path_exp = self.sub_stuff['expected_path'].format(**self.sub_stuff)
+        stderr_exp = self.sub_stuff['expected_stderr'].format(**self.sub_stuff)
```
Comment (cevich) [inline subtests/docker_cli/run_cgroups/cgroup_parent.py]: This will make expanding/extending the test later (if necessary) a lot easier to do.

File: subtests/docker_cli/run_cgroups/cgroup_parent.py [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+        for --cgroup-parent, and what to expect in stdout & stderr.
+        We replace the format strings {rand1} and {rand2} with two
+        pseudorandom strings, and {cid} with the container ID (once
+        we have it).
+        """
+        self.sub_stuff['rand1'] = utils.generate_random_string(8)
+        self.sub_stuff['rand2'] = utils.generate_random_string(8)
+        self.sub_stuff['cgroup_parent'] = cg_parent.format(**self.sub_stuff)
+        self.sub_stuff['expected_path'] = expected_path
+        self.sub_stuff['expected_stderr'] = expected_stderr
```
Comment (cevich) [inline subtests/docker_cli/run_cgroups/cgroup_parent.py]: BTW (in case you were wondering), test-data is kept in ``stuff`` and ``sub_stuff`` (they could be merged to use same name) is to keep the runtime namespace separated from the class and API.  It's just a simple way to avoid smashed toes (which has happened, along days spent tracking down ``-/_`` mix-ups and speling mistakes).

File: subtests/docker_cli/run_cgroups/run_cgroups.py (hunk 2 above)
Comment (cevich) [inline subtests/docker_cli/run_cgroups/run_cgroups.py]: Yet another idea: Just copy paste all the ``self.sub_stuff[blah] = blah`` lines into a separate ``initialize()`` call for each sub-subtest.  I realize it's duplication/repetition, however it's only a few lines, and it would allow one or more sub-subtests to diverge in the future w/o having to add all kinds of extra parameters/logic into the common ``_setup()`` method.  Anyway, something to consider because- if you're going to check/raise ``DockerTestNAError`` on too low a version, doing so inside ``initialize()`` may make more sense than inside ``_setup()`` (name wise).  But I'm not too particular on these things.  Just some ideas.

**Comment by cevich** [comment]
Uggg, I'm seriously mad this dumb 'Type less with saved replies' box keeps popping up.

Changes/additions look great, well written, easy to read and understand.  I shared some ideas, really the only thing I'd like to see for sure is the raising of ``DockerTestNAError`` when docker version < 1.10  (I triple checked this is what martin wanted, even lokesh asked him about it).

Nice job, especially for a hot-potato.

File: config_defaults/subtests/docker_cli/run_cgroups.ini (hunk 1 above)
Comment (edsantiago) [inline config_defaults/subtests/docker_cli/run_cgroups.ini]: Double ugh. 'docker version' returns 1.9.1, API Version 1.21, on both good and bad builds. Package version returns docker-1.9.1-25.el7.x86_64, but that won't help with Fedora/CentOS. Taking a step back now to let this simmer in my background brain.

File: subtests/docker_cli/run_cgroups/cgroup_parent.py (hunk 9 above)
Comment (edsantiago) [inline subtests/docker_cli/run_cgroups/cgroup_parent.py]: Thanks for the clarification on `stuff` vs `sub_stuff`! Any guideline on which one is preferred for future code?

File: subtests/docker_cli/run_cgroups/cgroup_parent.py (hunk 5 above)
Comment (edsantiago) [inline subtests/docker_cli/run_cgroups/cgroup_parent.py]: You might have noticed that the -D is also affecting CI results on Fedora. I don't know enough to have a strong opinion on keeping -D vs removing it, but my uninformed preference right now is to nuke it.

File: subtests/docker_cli/run_cgroups/run_cgroups.py (hunk 2 above)
Comment (edsantiago) [inline subtests/docker_cli/run_cgroups/run_cgroups.py]: My first approach was to overload `super.initialize`, passing those three extra args. I didn't feel comfortable with that because it changed the `initialize` signature and expectation. But would it be OK?

File: subtests/docker_cli/run_cgroups/cgroup_parent.py (hunk 3 above)
Comment (cevich) [inline subtests/docker_cli/run_cgroups/cgroup_parent.py]: Yep that's fine, I'm not too particular on the names as long as they make sense (which I fail at often).

File: config_defaults/subtests/docker_cli/run_cgroups.ini (hunk 1 above)
Comment (cevich) [inline config_defaults/subtests/docker_cli/run_cgroups.ini]: Just check and run test if >= '1.10' - those were our orders, and will help you manage your caffeine-quota :smile:

File: subtests/docker_cli/run_cgroups/cgroup_parent.py (hunk 9 above)
Comment (cevich) [inline subtests/docker_cli/run_cgroups/cgroup_parent.py]: My preference would be in the future (like for 1.9.1) we'd just edit ``dockertest.subtestbase`` and call them all 'stuff'.  There was a (now irrelevant) reason for the differentiation, now it's just a PITA.

File: subtests/docker_cli/run_cgroups/run_cgroups.py (hunk 2 above)
Comment (cevich) [inline subtests/docker_cli/run_cgroups/run_cgroups.py]: Ya, signature for those methods is burried deep and in non-obvious ways inside Autotest.  TBH, it's only like 5-7 items, I'd just copy-paste them into an overloaded ``initialize()`` w/o arguments.  That allows future modification to whomever's heart's content w/o any affect on other code.

Comment (cevich) [inline subtests/docker_cli/run_cgroups/run_cgroups.py]: ...splitting this all off into your own subtest is fine too, then you're not held to any old duct-tape previously applied to this ancient code.

Comment (edsantiago) [inline subtests/docker_cli/run_cgroups/run_cgroups.py]: Already split. (And, sidenote, I added a `cleanup` to remove stray cgroups). I will push tomorrow, not today, because my brain is shutting down. I'll also look into refactoring all the DockerVersion checks, because with this addition we'll have three and that's too many to duplicate.

File: subtests/docker_cli/run_cgroups/cgroup_parent.py (hunk 5 above)
Comment (cevich) [inline subtests/docker_cli/run_cgroups/cgroup_parent.py]: okay, opened #583

**Comment by edsantiago** [comment]
1.10 check now in place.

File: subtests/docker_cli/run_cgroup_parent/run_cgroup_parent.py [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+class run_cgroup_parent_path_with_hyphens(run_cgroup_parent_base):
+    """
+    Hyphens result in a more complicated path.
+    """
+
+    def initialize(self):
+        super(run_cgroup_parent_path_with_hyphens, self).initialize()
+        self._setup("{rand1}-{rand2}.slice")
+        self._expect(path="/{rand1}.slice/{rand1}-{rand2}.slice"
+                     "/docker-{cid}.scope")
```
Comment (cevich) [inline subtests/docker_cli/run_cgroup_parent/run_cgroup_parent.py]: hahaha, I had a feeling starting fresh would be a lot easier.  That other cgroups test is quite crusty.

**Comment by cevich** [comment]
Okay, merging this, it'll show up in the internal 'autobrew' CI builds, those pull the latest/greatest RPMs from everything (which is why it's always broken) :grinning:
